File: lxb/leetcode/combinations.cpp
http://oj.leetcode.com/problems/combinations/

class Solution {
public:
    vector<vector<int> > combine(int n, int k) {
        // Note: The Solution object is instantiated only once and is reused by each test case.
        vector<vector<int> > result;
        vector<int> oneSolution;
        
        combineHelper(k, n, oneSolution, result,1);
        return result;
    }
    
    void combineHelper(int k, int n, vector<int> s, vector<vector<int> > &r,int index){
        //base
        if (k==0){
            r.push_back(s);
            return;
        }
        
        if (index <= n)
        {
            //choose current one:
            s.push_back(index);
            combineHelper(k-1, n, s, r, index+1);
            
            //or not choose.
            s.pop_back();
            combineHelper(k, n, s, r, index+1);
        }
    }
};

File: lxb/leetcode/java/Combinations.java
//Focus: Runtime Error 1,1,
public class Solution {
    public ArrayList<ArrayList<Integer>> combine(int n, int k) {
        // Note: The Solution object is instantiated only once and is reused by each test case.
        
        ArrayList<Integer> oneSol = new ArrayList<Integer>();
        ArrayList<ArrayList<Integer>> res = new ArrayList<ArrayList<Integer>>();
        
        dfs(n,k,oneSol,res);
        return res;
    }
    
    private void dfs(int n, int k, ArrayList<Integer> oneSol, ArrayList<ArrayList<Integer>> res){
        //base
        if (k==0){
            res.add(oneSol);
            return;
        }
        
        if (n<k) return;
        
        // not choose n
        dfs(n-1, k, oneSol, res);
        
        // or choose n
        oneSol.add(n);
        dfs(n-1, k-1, oneSol, res);
        oneSol.remove(n);
    }
}
